src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/References/ReferenceCleanupService.cs

```diff
@@ -0,0 +1,252 @@
+
+        private readonly ConfiguredProject _configuredProject;
+        private readonly IVsUIService<DTE> _dte;
+        private readonly IVsUIService<SVsSolution, IVsSolution> _solution;
+
+        private Dictionary<ReferenceType, string> _referenceTypes = new Dictionary<ReferenceType, string>();
+
+        [ImportingConstructor]
+        public ReferenceCleanupService(ConfiguredProject configuredProject, IVsUIService<SDTE, DTE> dte, IVsUIService<SVsSolution, IVsSolution> solution)
+        {
+            _configuredProject = configuredProject;
+            _dte = dte;
+            _solution = solution;
+        }
+
+        public Task<string> GetProjectAssetsFilePathAsync(string projectPath, CancellationToken cancellationToken)
+        {
+            throw new NotImplementedException();
+        }
+
+        public async Task<ImmutableArray<ProjectSystemReferenceInfo>> GetProjectReferencesAsync(string projectPath, string targetFrameworkMoniker, CancellationToken cancellationToken)
+        {
+            List<ProjectSystemReferenceInfo> references;
+
+            try
+            {
+                var activeConfiguredProject = await GetActiveConfiguredProjectByPathAsync(projectPath);
+
+                references = await GetAllReferencesInConfiguredProjectAsync(activeConfiguredProject);
+            }
+            catch
+            {
+                throw new InvalidProjectFileException();
+            }
+
+            return references.ToImmutableArray();
+        }
+
+        private async Task<ConfiguredProject> GetActiveConfiguredProjectByPathAsync(string projectPath)
+        {
+            var unconfiguredProjectPath = FindUnconfiguredProjectByPath(projectPath);
+
+            var activeConfiguredProject = await unconfiguredProjectPath.GetSuggestedConfiguredProjectAsync();
+            if (activeConfiguredProject is null)
+            {
+                throw new InvalidProjectFileException();
+            }
+
+            return activeConfiguredProject;
+        }
+
+        private UnconfiguredProject FindUnconfiguredProjectByPath(string projectPath)
+        {
+            var unconfigProjects = _configuredProject.Services.ProjectService.LoadedUnconfiguredProjects;
+
+            var unconfiguredProject = unconfigProjects.First(project =>
+                StringComparers.Paths.Equals((string)project.FullPath, projectPath));
+
+            return unconfiguredProject;
+        }
+
+        private async Task<List<ProjectSystemReferenceInfo>> GetAllReferencesInConfiguredProjectAsync(ConfiguredProject selectedConfiguredProject)
+        {
+            GetProjectSnapshotAsync(selectedConfiguredProject);
+
+            var references = GetReferences();
+
+            return references;
+        }
+
+        private void GetProjectSnapshotAsync(ConfiguredProject selectedConfiguredProject)
+        {
+            _projectReferenceHandler.GetProjectSnapshot(selectedConfiguredProject);
```

Comment:
This pattern would be much nicer if this was a dictionary of reference type -> to the handler, and you just walked the dictionary instead of hanving "_projectReferenceHandler", "_packageReferenceHandler", etc. Then adding a new reference type is a 1 line change to this class instead of changing all usages.

Reply:
done